src/probability/martingale/borel_cantelli.lean

```diff
@@ -0,0 +1,407 @@
+In particular, we define
+$$
+  f_n := \sum_{k < n} \mathbf{1}_{s_{n + 1}} - \mathbb{P}[s_{n + 1} \mid \mathcal{F}_n].
+$$
+Then, as a martingale is both a sub and a super-martingale, the set for which it is unbounded from
+above must agree with the set for which it is unbounded from below almost everywhere. Thus, it
+can only converge to $\pm \infty$ with probability 0. Thus, by considering
+$$
+  \limsup_n s_n = \{\sum_n \mathbf{1}_{s_n} = \infty\}
+$$
+almost everywhere, the result follows.
+-/
+
+lemma martingale.bdd_above_range_iff_bdd_below_range [is_finite_measure μ]
+  (hf : martingale f ℱ μ) (hbdd : ∀ᵐ ω ∂μ, ∀ i, |f (i + 1) ω - f i ω| ≤ R) :
+  ∀ᵐ ω ∂μ, bdd_above (set.range (λ n, f n ω)) ↔ bdd_below (set.range (λ n, f n ω)) :=
+begin
+  have hbdd' : ∀ᵐ ω ∂μ, ∀ i, |(-f) (i + 1) ω - (-f) i ω| ≤ R,
+  { filter_upwards [hbdd] with ω hω i,
+    erw [← abs_neg, neg_sub, sub_neg_eq_add, neg_add_eq_sub],
+    exact hω i },
+  have hup := hf.submartingale.bdd_above_iff_exists_tendsto hbdd,
+  have hdown := hf.neg.submartingale.bdd_above_iff_exists_tendsto hbdd',
+  filter_upwards [hup, hdown] with ω hω₁ hω₂,
+  have : (∃ c, tendsto (λ n, f n ω) at_top (𝓝 c)) ↔ ∃ c, tendsto (λ n, (-f) n ω) at_top (𝓝 c),
+  { split; rintro ⟨c, hc⟩,
+    { exact ⟨-c, hc.neg⟩ },
+    { refine ⟨-c, _⟩,
+      convert hc.neg,
+      simp only [neg_neg, pi.neg_apply] } },
+  rw [hω₁, this, ← hω₂],
+  split; rintro ⟨c, hc⟩; refine ⟨-c, λ ω hω, _⟩,
+  { rw mem_upper_bounds at hc,
+    refine neg_le.2 (hc _ _),
+    simpa only [pi.neg_apply, set.mem_range, neg_inj] },
+  { rw mem_lower_bounds at hc,
+    simp_rw [set.mem_range, pi.neg_apply, neg_eq_iff_neg_eq, eq_comm] at hω,
+    refine le_neg.1 (hc _ _),
+    simpa only [set.mem_range] }
+end
+
+lemma martingale.ae_not_tendsto_at_top_at_top [is_finite_measure μ]
+  (hf : martingale f ℱ μ) (hbdd : ∀ᵐ ω ∂μ, ∀ i, |f (i + 1) ω - f i ω| ≤ R) :
+  ∀ᵐ ω ∂μ, ¬ tendsto (λ n, f n ω) at_top at_top :=
+by filter_upwards [hf.bdd_above_range_iff_bdd_below_range hbdd] with ω hω htop using
+    unbounded_of_tendsto_at_top htop (hω.2 $ bdd_below_range_of_tendsto_at_top_at_top htop)
+
+lemma martingale.ae_not_tendsto_at_top_at_bot [is_finite_measure μ]
+  (hf : martingale f ℱ μ) (hbdd : ∀ᵐ ω ∂μ, ∀ i, |f (i + 1) ω - f i ω| ≤ R) :
+  ∀ᵐ ω ∂μ, ¬ tendsto (λ n, f n ω) at_top at_bot :=
+by filter_upwards [hf.bdd_above_range_iff_bdd_below_range hbdd] with ω hω htop using
+    unbounded_of_tendsto_at_bot htop (hω.1 $ bdd_above_range_of_tendsto_at_top_at_bot htop)
+
+namespace borel_cantelli
+
+/-- Auxiliary definition required to prove Lévy's generalization of the Borel-Cantelli lemmas.
+The sum of the differences of the indicator functions with their conditional expectation forms a
+martingale satisfying the conditions of the one sided martingale bound. -/
+noncomputable
+def mgale (ℱ : filtration ℕ m0) (μ : measure Ω) (s : ℕ → set Ω) (n : ℕ) : Ω → ℝ :=
+∑ k in finset.range n, ((s (k + 1)).indicator 1 - μ[(s (k + 1)).indicator 1 | ℱ k])
```

Comment:
This object is what you get if you apply (the proof of) the centering lemma 9.10 in Foundations of Modern Probability (3rd ed.) to obtain a martingale from the process `∑ k in finset.range n, (s (k + 1)).indicator 1`.
Could you define two functions that take a process as input, and give as outputs the martingale M and the adapted process A defined in the centering lemma 9.10? Then prove some API for those.

Reply:
Update: I am currently doing this and will PR those soon.

Reply:
See #16517

Reply:
Thanks!